core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -0,0 +1,514 @@
+   * via {@link GzipMetadataReader}.
+   */
+  private final CompositeReadableBuffer gzippedData = new CompositeReadableBuffer();
+
+  private final CRC32 crc = new CRC32();
+
+  private final GzipMetadataReader gzipMetadataReader = new GzipMetadataReader();
+  private final byte[] inflaterInput = new byte[INFLATE_BUFFER_SIZE];
+  private int inflaterInputStart;
+  private int inflaterInputEnd;
+  private Inflater inflater;
+
+  private State state = State.HEADER_MAGIC;
+  private boolean closed = false;
+
+  /** Output buffer for inflated bytes. */
+  private byte[] inflaterOutput;
+
+  private int inflaterOutputEnd;
+
+  /** Extra state variables for parsing gzip header flags. */
+  private int gzipHeaderFlag;
+
+  private int headerExtraToRead;
+
+  /* Number of inflated bytes per gzip stream, used to validate the gzip trailer. */
+  private long expectedGzipTrailerIsize;
+
+  /** Tracks gzipped bytes consumed during each {@link #inflateBytes} call. */
+  private int bytesConsumed = 0;
+
+  /**
+   * Returns true when all of {@code gzippedData} has been input to the inflater and the inflater is
+   * unable to produce more output.
+   */
+  boolean isStalled() {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    return gzippedData.readableBytes() == 0 && (inflaterInputStart == inflaterInputEnd);
+  }
+
+  /**
+   * Returns true when there is gzippedData that has not been input to the inflater or the inflater
+   * has not consumed all of its input.
+   */
+  boolean hasPartialData() {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    return gzippedData.readableBytes() != 0 || (inflaterInputStart < inflaterInputEnd);
+  }
+
+  /**
+   * Adds more gzipped data, which will be consumed only when needed to fulfill requests made via
+   * {@link #inflateBytes}.
+   */
+  void addGzippedBytes(ReadableBuffer buffer) {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    gzippedData.addBuffer(buffer);
+  }
+
+  @Override
+  public void close() {
+    if (!closed) {
+      closed = true;
+      gzippedData.close();
+      if (inflater != null) {
+        inflater.end();
+        inflater = null;
+      }
+    }
+  }
+
+  /**
+   * Attempts to inflate up to {@code n} bytes of data into {@code bufferToWrite}. This method will
+   * always write as many inflated bytes as it can produce, up to a maximum of {@code
+   * bytesRequested}.
+   *
+   * <p>This method may consume gzipped bytes without writing any data to {@code bufferToWrite}, and
+   * may also write data to {@code bufferToWrite} without consuming additional gzipped bytes (if the
+   * inflater on an earlier call consumed the bytes necessary to produce output).
+   *
+   * @param n max number of bytes to inflate
+   * @param bufferToWrite destination for inflated data
+   * @return gzipped bytes consumed by the call (NOT the number of inflated bytes written)
+   */
+  int inflateBytes(int n, CompositeReadableBuffer bufferToWrite)
```

Comment:
I'd really prefer the output buffer to be passed in here. `(byte[], int, int)`-style like `InputStream.read()` and `inflater.inflate()`.

Reply:
In earlier discussion I mentioned just allocating the entire message. I was only thinking about mobile. For us to turn this on for all users, we probably do want to do the chunking, granted it could be at a much larger size (like 2 MB).

Reply:
Done - the chunking will be done by message deframer